src/ccmain/thresholder.cpp

```diff
@@ -184,6 +185,45 @@ void ImageThresholder::SetImage(const Image pix) {
   Init();
 }
 
+std::tuple<bool, Image, Image, Image> ImageThresholder::Threshold(
+                                                         ThreshMethod method) {
+  Image pix_grey = nullptr;
+  Image pix_binary = nullptr;
+  Image pix_thresholds = nullptr;
+
+  if (image_width_ > INT16_MAX || image_height_ > INT16_MAX) {
+    tprintf("Image too large: (%d, %d)\n", image_width_, image_height_);
+    return std::make_tuple(false, nullptr, nullptr, nullptr);
+  }
+  
+  if (pix_channels_ == 0) {
+    // We have a binary image, but it still has to be copied, as this API
+    // allows the caller to modify the output.
+    Image original = GetPixRect();
+    pix_binary = original.copy();
+    original.destroy();
+    return std::make_tuple(false, nullptr, pix_binary, nullptr);
+  }
+
+  pix_grey = GetPixRectGrey();
+
+  if (method == ThreshMethod::Otsu || method >= ThreshMethod::Count) {
+    method = ThreshMethod::AdaptiveOtsu;
+  }
+
+  int r;
+  if (method == ThreshMethod::AdaptiveOtsu) {
+    r = pixOtsuAdaptiveThreshold(pix_grey, 300, 300, 0, 0, 0.1, 
+                                 pix_thresholds.a(), pix_binary.a());
```

Comment:
@amitdo I would advice making the tile size depenent on the pixel density of the input (`estimated_res_`) or on the pixel resolution (`image_width_` and `image_height_`). How large our tiles are relatively has a big impact: Is it likely to have white-only segments? Or (on the other extreme) will segments stretch across most of the page (conflating shaded and bright areas)?

Reply:
Furthermore, I think it would be useful to expose the `scorefract=0.1` default as a tunable Tesseract parameter.

Reply:
#3577 adds configurable variables to control thresholding.